include/clap/ext/draft/context-menu.h

```diff
@@ -0,0 +1,121 @@
+} clap_context_menu_accelerator_key_t;
+
+// Describes an accelerator which can consist of a key strike combo
+typedef struct clap_context_menu_accelerator {
+   uint32_t combo_size;
+   clap_context_menu_accelerator_key_t keys[4];
+} clap_context_menu_accelerator_t;
+
+// Context menu builder
+typedef struct clap_context_menu_builder {
+   void *ctx;
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_entry)(const struct clap_context_menu_builder *builder,
+                             const char                             *label,
+                             const clap_context_menu_accelerator_t  *accelerator,
+                             bool                                    is_enabled,
+                             clap_id                                 action_id);
+
+   // accelerator may be null
+   bool(CLAP_ABI *add_checkbox)(const struct clap_context_menu_builder *builder,
+                                const char                             *label,
+                                const clap_context_menu_accelerator_t  *accelerator,
+                                bool                                    is_enabled,
+                                bool                                    is_checked,
+                                clap_id                                 action_id);
+
+   bool(CLAP_ABI *add_separator)(const struct clap_context_menu_builder *builder);
+
+   // Each begin_submenu() must have a corresponding end_submenu()
+   bool(CLAP_ABI *begin_submenu)(const struct clap_context_menu_builder *builder,
+                                 const char                             *label);
+
+   bool(CLAP_ABI *end_submenu)(const struct clap_context_menu_builder *builder);
+} clap_context_menu_builder_t;
+
+typedef struct clap_plugin_context_menu {
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *populate)(const clap_plugin_t               *plugin,
+                            const clap_context_menu_target_t  *target,
+                            const clap_context_menu_builder_t *builder);
+
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *perform)(const clap_plugin_t              *plugin,
+                           const clap_context_menu_target_t *target,
+                           clap_id                           action_id);
+} clap_plugin_context_menu_t;
+
+typedef struct clap_host_context_menu {
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *populate)(const clap_host_t                 *host,
+                            const clap_context_menu_target_t  *target,
+                            const clap_context_menu_builder_t *builder);
+
+   // If target is null, assume global context
+   // [main-thread]
+   bool(CLAP_ABI *perform)(const clap_host_t                *host,
+                           const clap_context_menu_target_t *target,
+                           clap_id action_id);
+
```

Comment:
Just for my understanding:

- If I understand correctly, calling populate and perform on the host from the plugin is for showing the plugin's own context menu but with (additional) host entries
- But through the clap_host_context_menu::popup method, it also could be the host's popup menu that shows up on top of the plugin. So clap_host_context_menu::populate and clap_host_context_menu::perform are a somehow redundant, aren't they?
- Or is it just so that we can have both options for context menus on top of the plugin?

Reply:
populate will add the host entries into the menu, so then the plugin can show its own menu, including the host entries.

perform is for executing an action after the menu item has been clicked.

If you call `host->popup()`, then the host will call `plugin->populate()` then show the menu then if a plugin menu item is clicked `plugin->perform()`